pandas/core/strings.py

```diff
@@ -1815,36 +1817,34 @@ def __init__(self, data):
     def _validate(data):
         from pandas.core.index import Index
 
-        if (isinstance(data, ABCSeries) and
-                not ((is_categorical_dtype(data.dtype) and
-                      is_object_dtype(data.values.categories)) or
-                     (is_object_dtype(data.dtype)))):
-            # it's neither a string series not a categorical series with
-            # strings inside the categories.
-            # this really should exclude all series with any non-string values
-            # (instead of test for object dtype), but that isn't practical for
-            # performance reasons until we have a str dtype (GH 9343)
+        if isinstance(data, ABCMultiIndex):
+                raise AttributeError('Can only use .str accessor with Index, '
+                                     'not MultiIndex')
+
+        # see src/inference.pyx which can contain string values
+        allowed_types = ['string', 'unicode', 'empty',
+                         'mixed', 'mixed-integer']
+        if isinstance(data, ABCSeries):
+            allowed_types = allowed_types + ['bytes']
+
+        values = data if isinstance(data, Index) else data.values
+        if is_categorical_dtype(data.dtype):
+            inf_type = lib.infer_dtype(values.categories)
+        else:
+            inf_type = lib.infer_dtype(values)
+
+        all_na_obj = is_object_dtype(values.dtype) and data.isna().all()
+
+        # same for Series and Index (that is not MultiIndex)
+        if inf_type not in allowed_types and not all_na_obj:
```

Comment:
rather than doing this here, I would for example have an if block that handles the empty (and checks if its object_dtype, then you immediately know if its ok or not)
then pass thru string and unicode. anything else is a raise.

Reply:
Ok in principle, but it's definitely necessary to allow `'mixed'`:
```
>>> lib.infer_dtype(pd.Series(['a', np.nan]))
'mixed'
```

Disallowing `'mixed-integer'` would also break existing code, but I mentioned in the OP:
> Finally, it would be possible to partially close #13877 by also excluding `'mixed-integer'` from the allowed inferred types.

Check out the example in #13877 - currently the `.str`-methods are generally permissive about just returning `NaN` if they can't compute individual values. Arguably, that's also the most desirable behaviour for things like `.str.startswith`. Could introduce an `'errors'`-kwarg where necessary, but all of that needs to happen on method level - the `.str`-constructor must allow these to pass.

Reply:
same comment as above, we need a very clear path where each case can be handled independently if need be.